### PR TITLE
Explicitly instantiate shape matrices

### PR DESCRIPTION
**Pull request by chleh**
Some rough measurements from envinf1 (`make -j4 && touch ProcessLib/GroundwaterFlow/GroundwaterFlowFEM.h && make -j4`). It looks like there is a bit of improvement, especially for non-dynamic shape matrices.

```
== building branch master
before initial make Thu Jun 16 11:16:13 CEST 2016

real	3m39.868s
user	8m55.049s
sys	1m19.031s
after initial make Thu Jun 16 11:19:53 CEST 2016
before touched make Thu Jun 16 11:19:53 CEST 2016

real	0m38.731s
user	0m32.065s
sys	0m7.518s
after touched make Thu Jun 16 11:20:31 CEST 2016
== building branch expl-inst-shape-mat
before initial make Thu Jun 16 11:20:34 CEST 2016

real	4m14.061s
user	8m44.046s
sys	1m18.185s
after initial make Thu Jun 16 11:24:48 CEST 2016
before touched make Thu Jun 16 11:24:48 CEST 2016

real	0m34.179s
user	0m28.239s
sys	0m7.685s
after touched make Thu Jun 16 11:25:22 CEST 2016
==== non-dynamic shape matrices
== building branch master
before initial make Thu Jun 16 11:25:27 CEST 2016

real	8m26.918s
user	13m31.169s
sys	2m52.399s
after initial make Thu Jun 16 11:33:54 CEST 2016
before touched make Thu Jun 16 11:33:54 CEST 2016

real	2m29.406s
user	1m42.570s
sys	1m23.979s
after touched make Thu Jun 16 11:36:23 CEST 2016
== building branch expl-inst-shape-mat
before initial make Thu Jun 16 11:36:38 CEST 2016

real	7m59.149s
user	12m13.200s
sys	2m38.705s
after initial make Thu Jun 16 11:44:37 CEST 2016
before touched make Thu Jun 16 11:44:37 CEST 2016

real	2m4.250s
user	1m19.624s
sys	1m9.896s
after touched make Thu Jun 16 11:46:41 CEST 2016
```

### CODE REVIEW

**Comment by norihiro-w** [comment]
thanks!

File: NumLib/Fem/CoordinatesMapping/NaturalCoordinatesMapping.cpp [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+    OGS_INSTANTIATE_NATURAL_COORDINATES_MAPPING_PART(                     \
+        RULE, SHAPE, 0, N, EigenDynamicShapeMatrixPolicy);              \
+    OGS_INSTANTIATE_NATURAL_COORDINATES_MAPPING_PART(                     \
+        RULE, SHAPE, 0, DNDR, EigenDynamicShapeMatrixPolicy);           \
+    OGS_INSTANTIATE_NATURAL_COORDINATES_MAPPING_PART(                     \
+        RULE, SHAPE, 0, N_J, EigenDynamicShapeMatrixPolicy);            \
+    OGS_INSTANTIATE_NATURAL_COORDINATES_MAPPING_PART(                     \
+        RULE, SHAPE, 0, DNDR_J, EigenDynamicShapeMatrixPolicy);         \
+    OGS_INSTANTIATE_NATURAL_COORDINATES_MAPPING_PART(                     \
+        RULE, SHAPE, 0, DNDX, EigenDynamicShapeMatrixPolicy)
```
Comment (endJunction) [inline NumLib/Fem/CoordinatesMapping/NaturalCoordinatesMapping.cpp]: formatting

File: NumLib/Fem/CoordinatesMapping/NaturalCoordinatesMapping.h [hunk 2]
```diff
@@ -41,23 +51,39 @@ struct NaturalCoordinatesMapping
...
+            T_SHAPE_FUNC,
+            T_SHAPE_MATRICES,
+            T_SHAPE_MATRIX_TYPE>(ele, natural_pt, shapemat);
+    }
 };
 
 } // NumLib
 
-#include "NaturalCoordinatesMapping-impl.h"
+// #include "NaturalCoordinatesMapping-impl.h"
```
Comment (endJunction) [inline NumLib/Fem/CoordinatesMapping/NaturalCoordinatesMapping.h]: rm

**Comment by chleh** [comment]
Jenkins, test this please.

**Comment by chleh** [comment]
"Slave went offline during the build". So windows on Jenkins seems to not work properly.

**Comment by norihiro-w** [comment]
:+1:

**Comment by norihiro-w** [comment]
isn't it needed to write 'extern template' in local assembly classes?

**Comment by chleh** [comment]
As far as I understood it, you have to write `extern template` if the template definition is visible to the code that uses it. However, here I'm hiding the definition and only have the declaration in the header file.
Of course, the compiler will here happily instantiate the template code, but it only sees the method declarations. The heavy stuff in the method definitions is hidden. I don't expect further compile time improvements from `extern template`, but I haven't measured it.

**Comment by norihiro-w** [comment]
i see. thank you for the explanation

**Comment by chleh** [comment]
Jenkins, test this please.

**Comment by chleh** [comment]
FYI: The normal windows Jenkins test went fine.

**Pull request merged**